packages/stream_video/lib/src/coordinator/open_api/coordinator_client_open_api.dart

```diff
@@ -76,26 +84,75 @@ class CoordinatorClientOpenApi extends CoordinatorClient {
   SharedEmitter<CoordinatorEvent> get events => _events;
   final _events = MutableSharedEmitterImpl<CoordinatorEvent>();
 
+  final _connectionState = MutableStateEmitterImpl<CoordinatorConnectionState>(
+    const CoordinatorDisconnected(),
+  );
+
   UserInfo? _user;
   CoordinatorWebSocketOpenApi? _ws;
   StreamSubscription<CoordinatorEvent>? _wsSubscription;
 
   @override
   Future<Result<None>> connectUser(UserInfo user) async {
-    _logger.d(() => '[connectUser] user: $user');
-    if (_user != null) {
-      _logger.w(() => '[connectUser] rejected (another user in use): $_user');
-      return Result.error(
-        'Another user is in use, please call "disconnectUser" first',
-      );
+    _logger.d(() => '[connectUser] user.id: ${user.id}');
+    final state = _connectionState.value;
+    if (state.isConnected) {
+      _logger.w(() => '[connectUser] rejected (already connected): $_user');
+      return const Result.success(none);
+    }
+    if (state.isConnecting) {
+      _logger.w(() => '[connectUser] wait (already connecting): $_user');
+      return _waitUntilConnected();
     }
+    _connectionState.value = CoordinatorConnectionState.connecting(
+      userId: user.id,
+    );
     _user = user;
     _ws = _createWebSocket(user).also((ws) {
-      _wsSubscription = ws.events.listen(_events.emit);
+      _wsSubscription = ws.events.listen((event) {
+        if (event is CoordinatorConnectedEvent) {
+          _logger.i(() => '[connectUser] WS connected');
+          _connectionState.value = CoordinatorConnectionState.connected(
+            userId: event.userId,
+          );
+        } else if (event is CoordinatorDisconnectedEvent) {
+          _logger.i(() => '[connectUser] WS disconnected');
+          _connectionState.value = CoordinatorConnectionState.disconnected(
+            userId: event.userId,
+            clientId: event.clientId,
+            closeCode: event.closeCode,
+            closeReason: event.closeReason,
+          );
+        }
+        _events.emit(event);
+      });
     });
     return openConnection();
   }
 
+  Future<Result<None>> _waitUntilConnected() async {
+    _logger.w(() => '[waitUntilConnected] user.id: ${_user?.id}');
+    return _connectionState
+        .firstWhere(
+      (it) => it.isConnected,
+      // TODO
+      // replace timeout with config value,
+      timeLimit: const Duration(seconds: 15),
+    )
+        .then((it) {
+      if (it.isConnected) {
+        _logger.v(() => '[waitUntilConnected] completed: $it');
+        return const Result.success(none);
+      } else {
+        _logger.e(() => '[waitUntilConnected] failed: $it');
+        return Result<None>.error('waitUntilConnected failed');
+      }
```

Comment:
We are only listening for `it.isConnected` then will there be any case it can be false?

Reply:
Good catch